src/pointer/ptr.rs

```diff
@@ -0,0 +1,1010 @@
+            //
+            // 2. It must be “dereferenceable” in the sense defined in the
+            //    module documentation; i.e.:
+            //
+            //    > The memory range of the given size starting at the pointer
+            //    > must all be within the bounds of a single allocated object.
+            //    > [2]
+            //
+            //   This is ensured by contract on all `Ptr`s.
+            //
+            // 3. The pointer must point to an initialized instance of `T`. This
+            //    is ensured by-contract on `Ptr`, because the
+            //    `VALIDITY_INVARIANT` is `Valid`.
+            //
+            // 4. You must enforce Rust’s aliasing rules. This is ensured by
+            //    contract on `Ptr`, because the `ALIASING_INVARIANT` is
+            //    `Shared`.
+            //
+            // [1]: https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.as_ref
+            // [2]: https://doc.rust-lang.org/std/ptr/index.html#safety
+            unsafe { raw.as_ref() }
+        }
+    }
+}
+
+/// State transitions between invariants.
+mod _transitions {
+    use super::*;
+
+    impl<
+            'a,
+            T,
+            const ALIASING_INVARIANT: invariant::Aliasing,
+            const ALIGNMENT_INVARIANT: invariant::Alignment,
+            const VALIDITY_INVARIANT: invariant::Validity,
+        > Ptr<'a, T, ALIASING_INVARIANT, ALIGNMENT_INVARIANT, VALIDITY_INVARIANT>
+    where
+        T: 'a + ?Sized,
+    {
+        /// Assumes that `Ptr`'s referent is validly-aligned for `T`.
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that `Ptr`'s referent conforms to the alignment
+        /// invariants of `T`.
+        #[inline]
+        pub(crate) const unsafe fn assume_aligned(
+            self,
+        ) -> Ptr<'a, T, ALIASING_INVARIANT, { invariant::Aligned }, VALIDITY_INVARIANT> {
+            // SAFETY: The caller promises that `self`'s referent is
+            // well-aligned for `T`.
+            unsafe { Ptr::from_ptr(self) }
+        }
+
+        /// Assumes that `Ptr`'s referent is as-initialized as `T`.
+        ///
+        /// # Safety
+        ///
+        /// The caller promises that `Ptr`'s referent only contains
+        /// uninitialized bytes in a subset of the uninitialized ranges in `T`.
+        ///
+        /// In other words, the `Ptr` is not required to refer to a valid `T`.
+        /// However, it must satisfy the requirement that uninitialized bytes
+        /// may only be present where it is possible for them to be present in
+        /// `T`. This is a dynamic property: if, at a particular byte offset, a
+        /// valid enum discriminant is set, the subsequent bytes may only have
+        /// uninitialized bytes as specificed by the corresponding enum.
+        ///
+        /// Formally, given `len = size_of_val_raw(candidate)`, at every byte
```

Comment:
```suggestion
        /// Formally, given `len = size_of_val_raw(self)`, at every byte
```

In general, update the rest of this comment to make sure it's consistent with the names used here.